tests/eas/st_hold.py

```diff
@@ -0,0 +1,408 @@
+
+test_conf = { "modules": ["bl", "cpufreq", "cgroups" ],
+              "tools": ["rt-app"],
+              "ftrace" : {
+                  "events" : ["cpu_frequency", "sched_switch", "sched_boost_cpu"],
+                  "buffsize" : 10240
+              },
+              "flags" : [ "ftrace" ]
+            }
+
+BOOSTED_CGROUP_NAME = '/boosted'
+UNBOOSTED_CGROUP_NAME = '/unboosted'
+FREQ_MARGIN = 20000
+TEST_CPU = 0
+
+class _SchedtuneHoldTest(unittest.TestCase):
+    """
+    Base class for all Schedtune-Hold feature tests. Contains
+    setup code and useful test methods.
+    When a derived class is instantiated, this sequence of events
+    happens:
+      _SchedtuneHoldTest:setUpClass called
+          -> connect target object
+          -> call derived_class:child_check_config() if implemented
+          -> call derived_class:populate_params() (must be implemented)
+          -> run workload
+          -> build Trace object in self.trace_obj
+    Then the actual test function is executed on the collected trace.
+
+    child_check_config is used for the two tests present where only one
+    can be run - which one is OK depends upon the status of the
+    SCHEDTUNE_BOOST_HOLD_ALL sched_feature. This function should raise a
+    unittest.SkipTest exception if it wants to prevent a derived class test
+    from running without causing a failure.
+    """
+    boosted_cgroup_name = BOOSTED_CGROUP_NAME
+    unboosted_cgroup_name = UNBOOSTED_CGROUP_NAME
+    frequency_margin = FREQ_MARGIN
+    test_cpu = TEST_CPU
+
+    @classmethod
+    def setUpClass(self):
+        self._log = logging.getLogger('StHoldTest')
+        self.params = {}
+        # connect the target
+        self.env = TestEnv(test_conf=test_conf)
+        self.trace_file = os.path.join(self.env.res_dir, '{}_trace.dat'.format(self.task_name))
+        self.target = self.env.target
+        # check that the SCHEDTUNE_HOLD_ALL feature exists in the target
+        self.check_hold_available()
+        # check any per-class requirements
+        self.child_check_config()
+        self.rta_name = 'rt-test-{}'.format(self.task_name.replace('/',''))
+        # ask derived class to populate workload config etc.
+        self.populate_params()
+        # ensure required cgroups exist
+        self.setup_cgroups()
+        # run workload
+        with self.env.freeze_userspace():
+            self.run_workload()
+        # collect and parse the trace
+        self.ftrace_obj = FTrace(self.trace_file)
+
+    @classmethod
+    def child_check_config(self):
+        pass
+
+    @classmethod
+    def run_workload(self):
+        self.env.ftrace.start()
+        self.rtapp.run(out_dir=self.env.res_dir, cgroup=self.cgroup_name, as_root=True)
+        self.env.ftrace.stop()
+        self.env.ftrace.get_trace(self.trace_file)
+        self.env.platform_dump(self.env.res_dir)
+
+    @classmethod
+    def setup_cpuset_cgroup(self, cgroup, cpus, mems):
```

Comment:
First parameter of class methods is usually called "cls" rather than self by convention

Reply:
ok :)